meilisearch-core/src/database.rs

```diff
@@ -319,23 +339,69 @@ impl Database {
         self.update_fn.swap(None);
     }
 
-    pub fn main_read_txn(&self) -> heed::Result<heed::RoTxn<MainT>> {
-        self.env.typed_read_txn::<MainT>()
+    pub fn main_read_txn(&self) -> MResult<MainReader> {
+        Ok(self.env.typed_read_txn::<MainT>()?)
+    }
+
+    pub(crate) fn main_write_txn(&self) -> MResult<MainWriter> {
+        Ok(self.env.typed_write_txn::<MainT>()?)
     }
 
-    pub fn main_write_txn(&self) -> heed::Result<heed::RwTxn<MainT>> {
-        self.env.typed_write_txn::<MainT>()
+    /// Calls f providing it with a writer to the main database. After f is called, makes sure the
+    /// transaction is commited. Returns whatever result f returns.
+    pub fn main_write<F, R, E>(&self, f: F) -> Result<R, E>
+    where
+        F: FnOnce(&mut MainWriter) -> Result<R, E>,
```

Comment:
Would probably be better to not restrict the user and [accept `FnMut`s](https://doc.rust-lang.org/core/ops/trait.FnMut.html).